src/bls12_381/ec.rs

```diff
@@ -1557,20 +1557,70 @@ pub mod g2 {
             }
         }
 
-        fn scale_by_cofactor(&self) -> G2 {
-            // G2 cofactor = (x^8 - 4 x^7 + 5 x^6) - (4 x^4 + 6 x^3 - 4 x^2 - 4 x + 13) // 9
-            // 0x5d543a95414e7f1091d50792876a202cd91de4547085abaa68a205b2e5a7ddfa628f1cb4d9e82ef21537e293a6691ae1616ec6e786f0c70cf1c38e31c7238e5
-            let cofactor = BitIterator::new([
-                0xcf1c38e31c7238e5,
-                0x1616ec6e786f0c70,
-                0x21537e293a6691ae,
-                0xa628f1cb4d9e82ef,
-                0xa68a205b2e5a7ddf,
-                0xcd91de4547085aba,
-                0x91d50792876a202,
-                0x5d543a95414e7f1,
+
+        fn psi(&mut self) {
+            let u = Fq2 { c0: Fq::one(), c1: Fq::one() };
+
+            let wy = u.pow(&[
+                0xdcff7fffffffd555,
+                0x0f55ffff58a9ffff,
+                0xb39869507b587b12,
+                0xb23ba5c279c2895f,
+                0x258dd3db21a5d66b,
+                0x0d0088f51cbff34d,
             ]);
-            self.mul_bits(cofactor)
+            let wy = wy.inverse().unwrap();
+
+
+            let wx = u.pow(&[
+                0x9354ffffffffe38e,
+                0x0a395554e5c6aaaa,
+                0xcd104635a790520c,
+                0xcc27c3d6fbd7063f,
+                0x190937e76bc3e447,
+                0x08ab05f8bdd54cde,
+            ]);
+            let wx = wx.inverse().unwrap();
```

Comment:
`wx` and `wy` could be precomputed. (I'd suggest still computing them this way, just in advance.)

Reply:
Please add a comment explaining where the exponents come from.